### PR TITLE
Make skipEmptyLines skip rows full of empty values

### PR DESCRIPTION
**Pull request by shamess**
Currently, skipEmptyLines only skips lines with no content but for the
new line character. Lines with nothing but the delimiters and the new
line character are just as useless, so skipping these too makes sense.

### CODE REVIEW

**Comment by shamess** [comment]
This is much faster than [my first attempt](https://github.com/mholt/PapaParse/pull/445), but the question about if this is too big a change for `skipEmptyLines` is still open.

**Comment by pokoli** [comment]
Indeed I think lines with empty values may be usefull for somebody. Let me put an example:

I have a survey questions results exported on csv and somebody responded all questions with an empty response. In that case I want the empty to row in order to count the response, but the empty value. 

I'm wondering why you need to remove the empty values on the parser. Could you explain your use case?

**Comment by shamess** [comment]
Our case is that we have customers who give us CSVs which are often exported from other systems. In the case of one customer they exported their data it added on 120k empty lines, like the kind that's being skipped here. In our UI, they're useless. It was surprising that the `skipEmptyLines` config wasn't skipping this kind.

Your use case makes sense too though.

What do you think about adding another configuration variable, to leave it up to the library user? I struggled to think of a name for what this might be though.

**Comment by pokoli** [comment]
Yes, I think we should add another configuration variable. 

For the name what about skipNullLines ?

What I'm sure is that the new option should be documented. We hold the documentation on the gh-pages branch on the same repo. Did you mind creating another PR to update the docs also?

**Comment by shamess** [comment]
I've updated this with the suggested new configuration option.

I'm excited to hear your thoughts!

**Comment by shamess** [comment]
Whilst writing the documentation, I realised that maybe "skipLinesWithEmptyFields" might explain the feature better than "skipNullLines". What do you think, @pokoli?

**Comment by pokoli** [comment]
Isn't this setting a little bit long? Probably it can be reduced to `SkipEmptyFieldsLines`

File: README.md [hunk 1]
```diff
@@ -55,6 +55,8 @@ Tests
 
 Papa Parse is under test. Download this repository, run `npm install`, then `npm test` to run the tests.
 
+The continuous integration server runs these tests using Node 0.12. You should ensure your tests pass locally using that version of Node.
```
Comment (pokoli) [inline README.md]: Not for this PR. If you wan to add this note please add another PR.

File: papaparse.js [hunk 2]
```diff
@@ -878,7 +878,7 @@
 			_delimiterError = false;
 			if (!_config.delimiter)
 			{
-				var delimGuess = guessDelimiter(input, _config.newline, _config.skipEmptyLines);
+				var delimGuess = guessDelimiter(input, _config.newline);
```
Comment (pokoli) [inline papaparse.js]: Why you removing it? Probably better to pass the config object directly. I prefer to not depend too much on global variables.

File: tests/test-cases.js [hunk 3]
```diff
@@ -1046,6 +1055,36 @@ var PARSE_TESTS = [
...
+	},
+	{
+		description: "Skip null lines and empty lines work well with each other",
+		notes: "",
+		input: 'a,b\n,\n\nc,d',
+		config: { skipNullLines: true, skipEmptyLines: true },
+		expected: {
+			data: [['a', 'b'], ['c', 'd']],
+			errors: []
+		}
```
Comment (pokoli) [inline tests/test-cases.js]: Can we add a test to ensure that empty lines are preserved when skipEmptyLines is false but skipNullLlines is true? 

It will be great to test the skipNullLines combined with header option.

File: papaparse.js [hunk 4]
```diff
@@ -946,10 +946,13 @@
 				_delimiterError = false;
 			}
 
-			if (_config.skipEmptyLines)
+			if (_config.skipEmptyLines || _config.skipNullLines)
 			{
 				for (var i = 0; i < _results.data.length; i++)
-					if (_results.data[i].length === 1 && _results.data[i][0] === '')
+					if (
+						(_config.skipEmptyLines && isEmptyLine(_results.data[i])) ||
```
Comment (pokoli) [inline papaparse.js]: Why creating two functions? It won't be simpler to create a function skipLine which makes the computation of both cases? I don't think each case is to complex to have it's own function.

File: papaparse.js (hunk 2 above)
Comment (shamess) [reply papaparse.js]: Sure thing - I can add it back.

I removed it because I didn't think of _config as a global, but more of a property of the ParseHandle object.

File: papaparse.js (hunk 4 above)
Comment (shamess) [reply papaparse.js]: I'll change this. I was following some advice that functions are free, so make a bunch of small, dumb ones. But I can see your point that they can sometimes get _too_ dumb.

Are you thinking the signature would look like this?

`skipLines(_results.data[i], _config.skipEmptyLines, _config.skipNullLines)`?

File: tests/test-cases.js (hunk 3 above)
Comment (shamess) [reply tests/test-cases.js]: Ah. That test was vital! `skipNullLines` is frustratingly also doing the job of `skipEmptyLines`. I'm figuring out a way to differentiate them now.

Comment (shamess) [reply tests/test-cases.js]: Unfortunately, this might not be easy to fix. By the time the data gets to `isEmptyLine`, for an emptyLine and a emptyFieldsLine the data structure looks the same.

For now, my company may just have to stick with the forked version of this library (or live with a few spreadsheets looking odd). I'll add the failing test I have, in case you've any ideas!

Comment (pokoli) [reply tests/test-cases.js]: Not sure if there is a use case for that, just wanted to see what's the behaviour of both flags together. Probably we can keep this behaviour but we should add a warning on the docs that explains that skipNullLines implies also skipEmptyLines

**Comment by pokoli** [comment]
Test suite is failing, so please make it pass otherwise we can not merge this changes.

**Comment by MonkeyDZeke** [comment]
This behavior is covered by the 'greedy' value passed to `skipEmptyLines` now: https://github.com/mholt/PapaParse/pull/544
